appengine/gophers/gophers-6/main.go

```diff
@@ -0,0 +1,271 @@
+		return "", err
+	}
+	if err := w.Close(); err != nil {
+		return "", err
+	}
+
+	const publicURL = "https://storage.googleapis.com/%s/%s"
+	return fmt.Sprintf(publicURL, storageBucketName, name), nil
+}
+
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path != "/" {
+		http.Redirect(w, r, "/", http.StatusFound)
+		return
+	}
+	ctx := appengine.NewContext(r)
+	params := templateParams{}
+
+	q := datastore.NewQuery("Post").Order("-Posted").Limit(20)
+	if _, err := q.GetAll(ctx, &params.Posts); err != nil {
+		log.Errorf(ctx, "Getting posts: %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		params.Notice = "Couldn't get latest posts. Refresh?"
+		indexTemplate.Execute(w, params)
+		return
+	}
+
+	if r.Method == "GET" {
+		indexTemplate.Execute(w, params)
+		return
+	}
+	// It's a POST request, so handle the form submission.
+
+	message := r.FormValue("message")
+	if message == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		params.Notice = "No message provided"
+		indexTemplate.Execute(w, params)
+		return
+	}
+
+	// Create a new Firebase App.
+	app, err := firebase.NewApp(ctx, &firebase.Config{
+		DatabaseURL:   "copy from Firebase Console > Overview > Add Firebase to your web app",
+		ProjectID:     "copy from Firebase Console > Overview > Add Firebase to your web app",
+		StorageBucket: "copy from Firebase Console > Overview > Add Firebase to your web app",
+	})
+	if err != nil {
+		params.Notice = "Couldn't authenticate. Try logging in again?"
+		params.Message = message // Preserve their message so they can try again.
+		indexTemplate.Execute(w, params)
+		return
+	}
+	// Create a new authenticator for the app.
+	auth, err := app.Auth(ctx)
+	if err != nil {
+		params.Notice = "Couldn't authenticate. Try logging in again?"
+		params.Message = message // Preserve their message so they can try again.
+		indexTemplate.Execute(w, params)
+		return
+	}
+	// Verify the token passed in by the user is valid.
+	tok, err := auth.VerifyIDTokenAndCheckRevoked(ctx, r.FormValue("token"))
+	if err != nil {
+		params.Notice = "Couldn't authenticate. Try logging in again?"
+		params.Message = message // Preserve their message so they can try again.
+		indexTemplate.Execute(w, params)
+		return
+	}
+	// Use the validated token to get the user's information.
+	user, err := auth.GetUser(ctx, tok.UID)
+	if err != nil {
+		params.Notice = "Couldn't authenticate. Try logging in again?"
+		params.Message = message // Preserve their message so they can try again.
+		indexTemplate.Execute(w, params)
+		return
+	}
+
+	post := Post{
+		UserID:  user.UID, // Include UserID in case Author isn't unique.
+		Author:  user.DisplayName,
+		Message: message,
+		Posted:  time.Now(),
+	}
+	params.Name = post.Author
+
+	// Get the image if there is one.
+	imageURL, err := uploadFileFromForm(ctx, r)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		params.Notice = "Error saving image: " + err.Error()
+		indexTemplate.Execute(w, params)
```

Comment:
I guess we should retain their message, too, not just their name?

Reply:
Done.